sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala

```diff
@@ -1535,6 +1539,211 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
       }
     }
   }
+
+  test("row index generation test") {
+    withTempPaths(2) { paths =>
+      paths.foreach(_.delete())
+      val df = (0 to 10000).toDF("id")
+      val rowIndexColName = RowIndexGenerator.ROW_INDEX_COLUMN_NAME
+
+      val pathWithNoRowIdx = paths.head.getAbsolutePath
+      val pathWithRowIdx = paths(1).getAbsolutePath
+
+      // no row index in schema.
+      df.repartition(1).write.format(dataSourceName).save(pathWithNoRowIdx)
+      val noRowIdxDF = spark.read.format(dataSourceName).load(pathWithNoRowIdx)
+      assert(!noRowIdxDF.columns.contains(rowIndexColName))
+
+      // With row index in schema.
+      val schemaWithRowIdx = df.schema.add(rowIndexColName, LongType, nullable = true)
+
+      df.repartition(1)
+        .write
+        .format(dataSourceName)
+        .option("parquet.block.size", 128) // Force the data to be split into multiple row groups.
+        .save(pathWithRowIdx)
+
+      val dfWithOnePartition = spark.read
+        .format(dataSourceName)
+        .schema(schemaWithRowIdx)
+        .load(pathWithRowIdx)
+      assert(dfWithOnePartition.filter(s"$rowIndexColName != id").count() == 0)
+    }
+  }
+
+  case class RowGroupStat(
+                           idx: Int,
+                           rowCnt: Long,
+                           offset: Long,
+                           byteSize: Long,
+                           isLast: Boolean
+                         )
+
+  def checkRowGroups(path: String, split: Option[(Long, Long)] = None): Seq[RowGroupStat] = {
+    val filter = split match {
+      case Some((start, end)) => ParquetMetadataConverter.range(start, end)
+      case None => NO_FILTER
+    }
+    val footer = ParquetFileReader.readFooter(
+      spark.sessionState.newHadoopConf(),
+      new Path(path),
+      filter)
+
+    /*
+     ParquetReadOptions options = HadoopReadOptions
+      .builder(configuration, file)
+      .withRange(split.getStart(), split.getStart() + split.getLength())
+      .build();
+     */
+    val blocks = footer.getBlocks.asScala
+    blocks.zipWithIndex.map { case (rowGroup, rowGroupIdx) =>
+      val rowCnt = rowGroup.getRowCount()
+      // Number of pages is convoluted to get.
+      val byteSize = rowGroup.getCompressedSize()
+      val offset = rowGroup.getRowIndexOffset()
+      val isLast = (rowGroupIdx + 1 == blocks.size)
+      RowGroupStat(rowGroupIdx, rowCnt, offset, byteSize, isLast)
+    }
+  }
```

Comment:
Dump information about each row group returned when reading the footer (with pushdown filter if needed). The third parameter is the row index offset.

Reply:
Test output: 
```
RowGroupStat(0,100,0,939,false)
RowGroupStat(1,100,100,949,false)
RowGroupStat(2,100,200,951,false)
RowGroupStat(3,100,300,950,false)
RowGroupStat(4,100,400,951,false)
RowGroupStat(5,100,500,946,false)
RowGroupStat(6,100,600,951,false)
RowGroupStat(7,100,700,952,false)
RowGroupStat(8,100,800,950,false)
RowGroupStat(9,100,900,952,false)
RowGroupStat(10,100,1000,946,false)
RowGroupStat(11,100,1100,944,false)
RowGroupStat(12,100,1200,952,false)
RowGroupStat(13,100,1300,950,false)
RowGroupStat(14,100,1400,950,false)
RowGroupStat(15,100,1500,953,false)
RowGroupStat(16,100,1600,944,false)
RowGroupStat(17,100,1700,953,false)
RowGroupStat(18,100,1800,950,false)
RowGroupStat(19,100,1900,950,false)
RowGroupStat(20,100,2000,951,false)
RowGroupStat(21,100,2100,944,false)
RowGroupStat(22,100,2200,944,false)
RowGroupStat(23,100,2300,951,false)
RowGroupStat(24,100,2400,950,false)
RowGroupStat(25,100,2500,952,false)
RowGroupStat(26,100,2600,951,false)
RowGroupStat(27,100,2700,944,false)
RowGroupStat(28,100,2800,954,false)
RowGroupStat(29,100,2900,952,false)
RowGroupStat(30,100,3000,952,false)
RowGroupStat(31,100,3100,950,false)
RowGroupStat(32,100,3200,944,false)
RowGroupStat(33,100,3300,946,false)
RowGroupStat(34,100,3400,950,false)
RowGroupStat(35,100,3500,952,false)
RowGroupStat(36,100,3600,950,false)
RowGroupStat(37,100,3700,950,false)
RowGroupStat(38,100,3800,945,false)
RowGroupStat(39,100,3900,953,false)
RowGroupStat(40,100,4000,952,false)
RowGroupStat(41,100,4100,950,false)
RowGroupStat(42,100,4200,950,false)
RowGroupStat(43,100,4300,946,false)
RowGroupStat(44,100,4400,944,false)
RowGroupStat(45,100,4500,950,false)
RowGroupStat(46,100,4600,952,false)
RowGroupStat(47,100,4700,952,false)
RowGroupStat(48,100,4800,953,false)
RowGroupStat(49,100,4900,944,false)
RowGroupStat(50,100,5000,954,false)
RowGroupStat(51,100,5100,952,false)
RowGroupStat(52,100,5200,950,false)
RowGroupStat(53,100,5300,952,false)
RowGroupStat(54,100,5400,943,false)
RowGroupStat(55,100,5500,943,false)
RowGroupStat(56,100,5600,952,false)
RowGroupStat(57,100,5700,952,false)
RowGroupStat(58,100,5800,952,false)
RowGroupStat(59,100,5900,953,false)
RowGroupStat(60,100,6000,944,false)
RowGroupStat(61,100,6100,953,false)
RowGroupStat(62,100,6200,950,false)
RowGroupStat(63,100,6300,950,false)
RowGroupStat(64,100,6400,950,false)
RowGroupStat(65,100,6500,944,false)
RowGroupStat(66,100,6600,946,false)
RowGroupStat(67,100,6700,950,false)
RowGroupStat(68,100,6800,952,false)
RowGroupStat(69,100,6900,952,false)
RowGroupStat(70,100,7000,953,false)
RowGroupStat(71,100,7100,946,false)
RowGroupStat(72,100,7200,951,false)
RowGroupStat(73,100,7300,950,false)
RowGroupStat(74,100,7400,952,false)
RowGroupStat(75,100,7500,951,false)
RowGroupStat(76,100,7600,946,false)
RowGroupStat(77,100,7700,944,false)
RowGroupStat(78,100,7800,950,false)
RowGroupStat(79,100,7900,952,false)
RowGroupStat(80,100,8000,952,false)
RowGroupStat(81,100,8100,953,false)
RowGroupStat(82,100,8200,944,false)
RowGroupStat(83,100,8300,951,false)
RowGroupStat(84,100,8400,952,false)
RowGroupStat(85,100,8500,949,false)
RowGroupStat(86,100,8600,952,false)
RowGroupStat(87,100,8700,946,false)
RowGroupStat(88,100,8800,944,false)
RowGroupStat(89,100,8900,952,false)
RowGroupStat(90,100,9000,951,false)
RowGroupStat(91,100,9100,950,false)
RowGroupStat(92,100,9200,953,false)
RowGroupStat(93,100,9300,944,false)
RowGroupStat(94,100,9400,954,false)
RowGroupStat(95,100,9500,950,false)
RowGroupStat(96,100,9600,949,false)
RowGroupStat(97,100,9700,952,false)
RowGroupStat(98,100,9800,944,false)
RowGroupStat(99,100,9900,947,true)
filtered:
RowGroupStat(0,100,0,950,false)
RowGroupStat(1,100,100,952,false)
RowGroupStat(2,100,200,951,false)
RowGroupStat(3,100,300,946,false)
RowGroupStat(4,100,400,944,false)
RowGroupStat(5,100,500,950,false)
RowGroupStat(6,100,600,952,false)
RowGroupStat(7,100,700,952,false)
RowGroupStat(8,100,800,953,false)
RowGroupStat(9,100,900,944,false)
RowGroupStat(10,100,1000,951,false)
RowGroupStat(11,100,1100,952,false)
RowGroupStat(12,100,1200,949,false)
RowGroupStat(13,100,1300,952,false)
RowGroupStat(14,100,1400,946,false)
RowGroupStat(15,100,1500,944,false)
RowGroupStat(16,100,1600,952,false)
RowGroupStat(17,100,1700,951,false)
RowGroupStat(18,100,1800,950,false)
RowGroupStat(19,100,1900,953,false)
RowGroupStat(20,100,2000,944,false)
RowGroupStat(21,100,2100,954,false)
RowGroupStat(22,100,2200,950,false)
RowGroupStat(23,100,2300,949,false)
RowGroupStat(24,100,2400,952,false)
RowGroupStat(25,100,2500,944,false)
RowGroupStat(26,100,2600,947,true)
```

Notice: The "filtered" blocks are the tail of the non-filtered blocks, but their row counts start from 0. Should be starting from 7300.